File: android/app/src/main/java/io/github/qwertyuiop1995/dsmnativeclient/domain/MutationResult.kt
package io.github.qwertyuiop1995.dsmnativeclient.domain

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
enum class MutationResultStatus {
    @SerialName("confirmedSuccess")
    CONFIRMED_SUCCESS,

    @SerialName("confirmedFailure")
    CONFIRMED_FAILURE,

    @SerialName("submittedButUnverified")
    SUBMITTED_BUT_UNVERIFIED,

    @SerialName("partialSuccess")
    PARTIAL_SUCCESS,

    @SerialName("cancelledBeforeSubmission")
    CANCELLED_BEFORE_SUBMISSION,

    @SerialName("cancellationRequestedAfterSubmission")
    CANCELLATION_REQUESTED_AFTER_SUBMISSION,

    @SerialName("permissionDenied")
    PERMISSION_DENIED,

    @SerialName("unsupported")
    UNSUPPORTED,
}

@Serializable
enum class MutationErrorCategory {
    @SerialName("validation")
    VALIDATION,

    @SerialName("authentication")
    AUTHENTICATION,

    @SerialName("permission")
    PERMISSION,

    @SerialName("conflict")
    CONFLICT,

    @SerialName("network")
    NETWORK,

    @SerialName("server")
    SERVER,

    @SerialName("unsupported")
    UNSUPPORTED,

    @SerialName("unknown")
    UNKNOWN,
}

@Serializable
data class MutationResultCounts(
    val succeeded: Int,
    val failed: Int,
    val unknown: Int,
) {
    init {
        require(succeeded >= 0 && failed >= 0 && unknown >= 0) {
            "mutation.invalid_count"
        }
    }
}

@Serializable
data class MutationResult(
    val schemaVersion: Int,
    val status: MutationResultStatus,
    val operation: String,
    val submitted: Boolean,
    val requiresRefresh: Boolean,
    val counts: MutationResultCounts,
    val errorCategory: MutationErrorCategory? = null,
    val localizationKey: String? = null,
    val diagnosticTag: String? = null,
) {
    init {
        require(schemaVersion == 1) { "mutation.unsupported_schema" }
        require(OPERATION_PATTERN.matches(operation)) { "mutation.invalid_operation" }
        require(localizationKey == null || SAFE_TAG_PATTERN.matches(localizationKey)) {
            "mutation.invalid_localization_key"
        }
        require(diagnosticTag == null || SAFE_TAG_PATTERN.matches(diagnosticTag)) {
            "mutation.invalid_diagnostic_tag"
        }
        validateState()
    }

    private fun validateState() {
        when (status) {
            MutationResultStatus.CONFIRMED_SUCCESS ->
                require(submitted && counts.failed == 0 && counts.unknown == 0) {
                    "mutation.inconsistent_success"
                }

            MutationResultStatus.CANCELLED_BEFORE_SUBMISSION ->
                require(
                    !submitted &&
                        !requiresRefresh &&
                        counts.succeeded == 0 &&
                        counts.failed == 0 &&
                        counts.unknown == 0,
                ) {
                    "mutation.inconsistent_pre_submission_cancellation"
                }

            MutationResultStatus.SUBMITTED_BUT_UNVERIFIED,
            MutationResultStatus.CANCELLATION_REQUESTED_AFTER_SUBMISSION,
            ->
                require(submitted && requiresRefresh) {
                    "mutation.unverified_requires_refresh"
                }

            MutationResultStatus.PARTIAL_SUCCESS ->
                require(
                    submitted &&
                        counts.succeeded > 0 &&
                        counts.failed + counts.unknown > 0,
                ) {
                    "mutation.invalid_partial_success"
                }

            MutationResultStatus.CONFIRMED_FAILURE,
            MutationResultStatus.PERMISSION_DENIED,
            MutationResultStatus.UNSUPPORTED,
            -> Unit
        }
    }

    private companion object {
        val OPERATION_PATTERN = Regex("^[a-z][A-Za-z0-9]*$")
        val SAFE_TAG_PATTERN = Regex("^[a-z0-9][a-z0-9._-]*$")
    }
}

File: windows/src/LanStash.Domain/MutationResult.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace LanStash.Domain;

[JsonConverter(typeof(MutationResultStatusJsonConverter))]
public enum MutationResultStatus
{
    ConfirmedSuccess,
    ConfirmedFailure,
    SubmittedButUnverified,
    PartialSuccess,
    CancelledBeforeSubmission,
    CancellationRequestedAfterSubmission,
    PermissionDenied,
    Unsupported,
}

[JsonConverter(typeof(MutationErrorCategoryJsonConverter))]
public enum MutationErrorCategory
{
    Validation,
    Authentication,
    Permission,
    Conflict,
    Network,
    Server,
    Unsupported,
    Unknown,
}

public sealed class MutationResultStatusJsonConverter()
    : JsonStringEnumConverter<MutationResultStatus>(
        JsonNamingPolicy.CamelCase,
        allowIntegerValues: false);

public sealed class MutationErrorCategoryJsonConverter()
    : JsonStringEnumConverter<MutationErrorCategory>(
        JsonNamingPolicy.CamelCase,
        allowIntegerValues: false);

public sealed record MutationResultCounts
{
    [JsonPropertyName("succeeded")]
    public int Succeeded { get; }

    [JsonPropertyName("failed")]
    public int Failed { get; }

    [JsonPropertyName("unknown")]
    public int Unknown { get; }

    [JsonConstructor]
    public MutationResultCounts(int succeeded, int failed, int unknown)
    {
        if (succeeded < 0 || failed < 0 || unknown < 0)
        {
            throw new ArgumentOutOfRangeException(nameof(succeeded), "mutation.invalid_count");
        }
        Succeeded = succeeded;
        Failed = failed;
        Unknown = unknown;
    }
}

public sealed record MutationResult
{
    [JsonPropertyName("schemaVersion")]
    public int SchemaVersion { get; }

    [JsonPropertyName("status")]
    public MutationResultStatus Status { get; }

    [JsonPropertyName("operation")]
    public string Operation { get; }

    [JsonPropertyName("submitted")]
    public bool Submitted { get; }

    [JsonPropertyName("requiresRefresh")]
    public bool RequiresRefresh { get; }

    [JsonPropertyName("counts")]
    public MutationResultCounts Counts { get; }

    [JsonPropertyName("errorCategory")]
    public MutationErrorCategory? ErrorCategory { get; }

    [JsonPropertyName("localizationKey")]
    public string? LocalizationKey { get; }

    [JsonPropertyName("diagnosticTag")]
    public string? DiagnosticTag { get; }

    [JsonConstructor]
    public MutationResult(
        int schemaVersion,
        MutationResultStatus status,
        string operation,
        bool submitted,
        bool requiresRefresh,
        MutationResultCounts counts,
        MutationErrorCategory? errorCategory = null,
        string? localizationKey = null,
        string? diagnosticTag = null)
    {
        if (schemaVersion != 1)
        {
            throw new ArgumentOutOfRangeException(
                nameof(schemaVersion),
                "mutation.unsupported_schema");
        }
        if (!IsValidOperation(operation))
        {
            throw new ArgumentException("mutation.invalid_operation", nameof(operation));
        }
        if (!IsValidSafeTag(localizationKey) || !IsValidSafeTag(diagnosticTag))
        {
            throw new ArgumentException("mutation.invalid_safe_tag");
        }
        ValidateState(status, submitted, requiresRefresh, counts);

        SchemaVersion = schemaVersion;
        Status = status;
        Operation = operation;
        Submitted = submitted;
        RequiresRefresh = requiresRefresh;
        Counts = counts;
        ErrorCategory = errorCategory;
        LocalizationKey = localizationKey;
        DiagnosticTag = diagnosticTag;
    }

    private static void ValidateState(
        MutationResultStatus status,
        bool submitted,
        bool requiresRefresh,
        MutationResultCounts counts)
    {
        var valid = status switch
        {
            MutationResultStatus.ConfirmedSuccess =>
                submitted && counts.Failed == 0 && counts.Unknown == 0,
            MutationResultStatus.CancelledBeforeSubmission =>
                !submitted &&
                !requiresRefresh &&
                counts.Succeeded == 0 &&
                counts.Failed == 0 &&
                counts.Unknown == 0,
            MutationResultStatus.SubmittedButUnverified or
                MutationResultStatus.CancellationRequestedAfterSubmission =>
                submitted && requiresRefresh,
            MutationResultStatus.PartialSuccess =>
                submitted &&
                counts.Succeeded > 0 &&
                counts.Failed + counts.Unknown > 0,
            _ => true,
        };
        if (!valid)
        {
            throw new ArgumentException("mutation.inconsistent_state");
        }
    }

    private static bool IsValidOperation(string? value)
    {
        return !string.IsNullOrEmpty(value) &&
            value[0] is >= 'a' and <= 'z' &&
            value.All(character =>
                character is >= 'a' and <= 'z' or
                    >= 'A' and <= 'Z' or
                    >= '0' and <= '9');
    }

    private static bool IsValidSafeTag(string? value)
    {
        return value is null ||
            (
                value.Length > 0 &&
                value.All(character =>
                    character is >= 'a' and <= 'z' or
                        >= '0' and <= '9' or
                        '.' or
                        '_' or
                        '-')
            );
    }
}
